### ISSUE TITLE
Adding orderby.case

### ISSUE THREAD

**Issue by Shaihuludus** [Adobe-Marketing-Cloud/asset-share-commons#253]
Hi, is there any way to add 
orderby.case=ignore to search order (both default search and sort component) (except extending components)?

I tried hidden parameter but it adds to group not directly under root. Maybe allowing user to choose if hidden parameter should or should not be added to new group

**Comment by davidjgonzalez** [comment]
I worry about breaking from the auto-grouping, since things can get very complicated very quickly (as evidenced by the group_3 issues, even WITH it, its complicated! :) ).. and there are few top-level predicates anyhow.

I think the easiest way to achieve this will be to add a checkbox to the Search Component's configuration, and then add it in the PagePredicateImpl...

Let us know if you want to take a stab at this and well make sure not to work on it (and instead work on other issues).

![image](https://user-images.githubusercontent.com/1451868/45031200-2c15d000-b01c-11e8-851d-3f6c4f3d24d1.png)

**Comment by davidjgonzalez** [comment]
For posterity - i mixed up the screenshot above; orderby's are not Group Parameters, but rather just predicates (like Type)

**Comment by davidjgonzalez** [comment]
@Shaihuludus i think this will have to be rolled back.

It appears that this flag wraps the sort by value in a `fn:lower-case(...)` in the underlying SQL2 query.

The problem with this, is it forces the sorting to be evaluated as a String sort, and sorts it lexicographically, and not numerically (im assuming same for dates)... for sorting Desc by Width, 900 is considered greater than 5000, since the first character of 9 is greater than 5.


The only hack in ASC i can think of is to add a checkbox for each Sort option specifying if it should be case sensitive or not, but that leaks a lot of impl details into the authoring experience which likely isnt good/safe.

We should get a fix in the product so this is handled cleanly - It would be good if someone could log an issue with Adobe Support for this.



```
query (predicate tree):
ROOT=group: limit=24, guessTotal=100, offset=0[
    {orderby=orderby: orderby=@jcr:content/metadata/tiff:ImageWidth, sort=desc, case=ignore}
    {type=type: type=dam:Asset}
    {2_group=group: or=true[
        {0_path=path: path=/content/dam/asset-share-commons/en/public}
        {1_path=path: path=/content/dam/we-retail/en}
    ]}
    {4_group=group: [
        {mainasset=mainasset: mainasset=true}
    ]}
    {5_group=group: [
        {property=property: property=jcr:content/contentFragment, operation=not}
    ]}
]
07.09.2018 08:38:24.986 *DEBUG* [0:0:0:0:0:0:0:1 [1536323904960] GET /content/asset-share-commons/en/light.results.html HTTP/1.1] com.day.cq.search.impl.builder.QueryImpl xpath query: (/jcr:root/content/dam/asset-share-commons/en/public//element(*, dam:Asset)[(not(jcr:content/@contentFragment))] | /jcr:root/content/dam/we-retail/en//element(*, dam:Asset)[(not(jcr:content/@contentFragment))]) order by fn:lower-case(jcr:content/metadata/@tiff:ImageWidth) descending
```

**Comment by sator1** [comment]
For strings fn:lower-case(string) does the trick.

This xpath query works fine:
/jcr:root/content/dam/a/b/c/d/e//element(*, dam:Asset)[(not(jcr:content/@contentFragment))] order by fn:lower-case(jcr:content/@cq:name) ascending

So there would have to some way to indicate with what kind of values query would be dealing with (like checkbox idea above).

Could this change go ahead to support string sorting or this still needs fix for lexicographic issue for number sorting  to be reported internally at oak end by Adobe Support. Discussed with Adobe Support Team.

**Comment by TonyG99** [comment]
For our use cases, the ability to correctly sort alphanumeric values (e.g. filename) and dates (e.g. last published date) is far more important than the ability to sort numeric values (e.g. pixel dimensions). I've been working in DAM for over a decade, and cannot think of many real-life use cases where users would want to sort by pixel dimensions.

BTW the CQ Daycare ticket we submitted for this issue is 160609 - Case sensitive sorting is misleading

**Comment by davidjgonzalez** [comment]
Thanks for submitting @TonyG99  ... FWIW, as an end-user of "DAMs", I find it very useful to sort by both filesize as well as dimensions; Often ill want to limit my searches to images that are of a (very)-hi res... and perhaps oddly, don't care about the filenames (or their sort order) at all. But, different strokes for different folks :)

I don't recall if the date sorting (can) break with its value being cast as as string, or if the "natural" way its serialized to a string allows the string sorting of the date representation to be interpreted accurately. im guessing time-zone differences would break it?

**Comment by TonyG99** [comment]
Yes, ideally it should be possible to sort numeric, alphanumeric, and date metadata values correctly, @davidjgonzalez.

**Comment by davidjgonzalez** [comment]
@TonyG99 ive been stewing on this, since it seems like we should be able to work around this (though i don't like that we have to work around it ;)) ... im just trying to think of a simple approach that doesn't create code-soup, and confuses users....

Ideally the Order By property selection would under the covers persist the data type, at which point it would be easy to check if order.case should be set or not... the problem if (i believe) the GraniteUI widget supports a label and value ... and im not super keen on jamming the type into the value and dealing w parsing things out... (ie. `<option value="jcr:content/metadata/foo@type=String">...`) .. im not sure if there's an elegant way we can pass that piece of metadata (the selected property type) in from the dialog submission.

I don't want to look up the value in AEM Publish on "render" (or query)... seems way overkill.

**Comment by Shaihuludus** [comment]
Hi, after long searching the best approach:
https://github.com/Adobe-Marketing-Cloud/asset-share-commons/pull/306

The solution allows author to select if sorting parameter should be treated with orderby.case or not. Every sort (default one or in sort component) has a checkbox with warning that it should not be checked if sort parameter is not text

**Comment by Shaihuludus** [comment]
I merged Yout pr into my branch and is working as it should. My pr should be have these changes and I think we are good to go :)

**Comment by davidjgonzalez** [comment]
Released in v1.6.10 - thanks @Shaihuludus !

**Comment by TonyG99** [comment]
Perfect timing -- saw this announcement just in time to announce it during the code handover presentation to our client! :-)

@davidjgonzalez , would you be interested in a short demo of what we've built using ASC? If so, shoot me an email: tony.gill@cognifide.com
